plugins/funind/recdef.ml

```diff
@@ -1615,6 +1615,23 @@ let com_eqn uctx nb_arg eq_name functional_ref f_ref terminate_ref
   in
   ()
 
+let check_relation_type env evd relation =
+  let err () = CErrors.user_err (strbrk "a well-founded relation is expected to be of type A->A->Prop for some type A.") in
+  let evd, t = Typing.type_of env evd relation in
+  let open Reductionops in
+  match EConstr.kind evd (whd_all env evd t) with
```

Comment:
The env below are incorrect since the Prod constructor binds a variable. Really, I think it'd be cleaner to use dest_arity and filter the result. (Also this is spaghetti code.)